lib/ddtrace/contrib/sinatra/tracer.rb

```diff
@@ -0,0 +1,165 @@
+          tracer: Datadog.tracer,
+          debug: false,
+          trace_agent_hostname: 'localhost',
+          trace_agent_port: 7777
+        }.freeze()
+
+        attr_accessor :cfg, :request_span
+
+        def initialize
+          @cfg = DEFAULT_CFG.dup()
+          @request_span = nil
+        end
+
+        def configure(**args)
+          args.each do |name, value|
+            self[name] = value
+          end
+
+          apply()
+        end
+
+        def apply
+          Datadog::Tracer.debug_logging = @cfg[:debug]
+
+          @cfg[:tracer].enabled = @cfg[:enabled]
+          @cfg[:tracer].configure(hostname: @cfg[:trace_agent_hostname],
+                                  port: @cfg[:trace_agent_port])
+        end
+
+        def [](key)
+          raise ArgumentError, "unknown setting '#{key}'" unless @cfg.key? key
+          @cfg[key]
+        end
+
+        def []=(key, value)
+          raise ArgumentError, "unknown setting '#{key}'" unless @cfg.key? key
+          @cfg[key] = value
+        end
+
+        def enabled?
+          @cfg[:enabled] && @cfg[:tracer]
+        end
+      end
+
+      # Datadog::Contrib::Sinatra::Tracer is a Sinatra extension which traces
+      # requests.
+      module Tracer
+        def route(verb, action, *)
+          # Keep track of the route name when the app is instanciated for an
+          # incoming request.
+          condition do
+            @datadog_route = action
+          end
+
+          super
+        end
+
+        # rubocop:disable Metrics/AbcSize
+        # rubocop:disable Metrics/CyclomaticComplexity
+        # rubocop:disable Metrics/MethodLength
+        # rubocop:disable Metrics/PerceivedComplexity
+        def self.registered(app)
+          ::Sinatra::Base.module_eval do
+            def render(engine, data, *)
+              cfg = settings.datadog_tracer
+
+              output = ''
+              if cfg.enabled?
+                tracer = cfg[:tracer]
+                tracer.trace('sinatra.render_template') do |span|
+                  span.set_tag('sinatra.template_engine', engine)
+                  span.set_tag('sinatra.template_name', data)
+                  output = super
+                end
+              else
+                output = super
+              end
+
+              output
+            end
+          end
+
+          app.set :datadog_tracer, TracerCfg.new()
+
+          app.configure do
+            app.settings.datadog_tracer.apply()
+          end
+
+          app.before do
+            cfg = settings.datadog_tracer
+            return unless cfg.enabled?
+
+            if @datadog_request_span
+              Datadog::Tracer.log.error('request span active in :before hook')
+              @datadog_request_span.finish()
+              @datadog_request_span = nil
+            end
+
+            tracer = cfg[:tracer]
+
+            span = tracer.trace('sinatra.request',
+                                service: cfg.cfg[:default_service],
```

Comment:
i think the colons might be specific to ruby 2.0. I _think_ we want to do `=>` here. it's a bit more portable. will defer to a ruby expert :)